### PR TITLE
Feature/swig node

### PR DESCRIPTION
**Pull request by luckychess**
### Description of the Change

Implementation of shared library for NodeJS.
Huge thanks to @haojun and @lasintez for help in implementing of this.

### Benefits

Possible to use Shared Model with NodeJS (And Electron)

### Possible Drawbacks 

None. But beware or package-lock.json - it sets exact versions of all dependencies; it's huge but it guarantees that ```npm install``` will work on any system.

### Usage Examples or Tests *[optional]*
See example/node folder.

### CODE REVIEW

**Comment by Warchant** [comment]
2750 :hocho:

**Comment by laSinteZ** [comment]
2,307 - example/node/package-lock.json
[Yes, we should commit it.](https://docs.npmjs.com/files/package-lock.json)

File: cmake/Modules/FindNodejs.cmake [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    HINTS
+    $ENV{NODE_DIR}
+    PATH_SUFFIXES bin
+    DOC "Node.js interpreter"
+)
+
+include (FindPackageHandleStandardArgs)
+
+# If compat-libuv package exists, it must be at start of include path
+find_path (UV_ROOT_DIR "uv.h" PATHS /usr/include/compat-libuv010 NO_DEFAULT_PATH)
```
Comment (Warchant) [inline cmake/Modules/FindNodejs.cmake]: what if there is no uv.h?

File: cmake/Modules/FindNodejs.cmake [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  add_include_dir(${NODE_ROOT_DIR}/include/src)
+  add_include_dir(${NODE_ROOT_DIR}/src)
+  add_include_dir(${NODE_ROOT_DIR}/include/node)
+  add_include_dir(${NODE_ROOT_DIR}/include/deps/v8/include)
+  add_include_dir(${NODE_ROOT_DIR}/deps/v8/include)
+  add_include_dir(${NODE_ROOT_DIR}/include/deps/uv/include)
+  add_include_dir(${NODE_ROOT_DIR}/deps/uv/include)
+else()
+  unset(NODEJS_INCLUDE_DIRS)
+  message(ERROR " - node.h not found")
```
Comment (Warchant) [inline cmake/Modules/FindNodejs.cmake]: there is no ERROR, only FATAL_ERROR

File: cmake/Modules/FindNodejs.cmake [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+endif()
+
+# Check that v8.h is in NODEJS_INCLUDE_DIRS
+find_path (V8_ROOT_DIR "v8.h" PATHS ${NODEJS_INCLUDE_DIRS})
+if (NOT V8_ROOT_DIR)
+  unset(NODEJS_INCLUDE_DIRS)
+  message(ERROR " - v8.h not found")
```
Comment (Warchant) [inline cmake/Modules/FindNodejs.cmake]: ERROR -> FATAL_ERROR

File: cmake/Modules/FindNodejs.cmake [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+endif()
+
+# Check that uv.h is in NODEJS_INCLUDE_DIRS
+find_path (UV_ROOT_DIR "uv.h" PATHS ${NODEJS_INCLUDE_DIRS})
+if (NOT UV_ROOT_DIR)
+  unset(NODEJS_INCLUDE_DIRS)
+  message(ERROR " - uv.h not found")
```
Comment (Warchant) [inline cmake/Modules/FindNodejs.cmake]: ERROR -> FATAL_ERROR

File: cmake/Modules/FindNodejs.cmake [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    else ()
+        set (NODE_VERSION_STRING "0.10.30")
+        set (NODE_VERSION_MAJOR "0")
+        set (NODE_VERSION_MINOR "10")
+        set (NODE_VERSION_PATCH "30")
+        set (V8_VERSION_MAJOR "3")
+        set (V8_VERSION_MINOR"14")
+        set (V8_VERSION_PATCH "5")
+        set (V8_VERSION_STRING "3.28.72")
+        message ("defaulted to node 0.10.30")
```
Comment (Warchant) [inline cmake/Modules/FindNodejs.cmake]: message(STATUS ...)

File: cmake/Modules/FindNodejs.cmake [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    endif ()
+    string (REGEX REPLACE "\n" "" NODE_VERSION_STRING ${NODE_VERSION_STRING})
+    string (REGEX REPLACE "\n" "" V8_VERSION_STRING ${V8_VERSION_STRING})
+    message ("INFO - Node version is " ${NODE_VERSION_STRING})
+    message ("INFO - Node using v8 " ${V8_VERSION_STRING})
```
Comment (Warchant) [inline cmake/Modules/FindNodejs.cmake]: message(STATUS ...) for both lines

File: example/node/package-lock.json [hunk 7]
```diff
@@ -0,0 +1,2307 @@
+{
```
Comment (Warchant) [inline example/node/package-lock.json]: gitignore this file

File: example/node/prepare.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+cmake --build build_cmake/ --target irohanode -- -j"$(getconf _NPROCESSORS_ONLN)";
+
+# npm install implicitly runs node-gyp rebuild
+npm install
+
+# Generating protobuff files
+mkdir pb
+mkdir pb/grpc
+protoc --proto_path=../../schema --js_out=import_style=commonjs,binary:./pb block.proto primitive.proto commands.proto queries.proto responses.proto endpoint.proto
+protoc --proto_path=../../schema --js_out=import_style=commonjs,binary:./pb/grpc --plugin=protoc-gen-grpc=./node_modules/grpc-tools/bin/grpc_node_plugin --grpc_out=./pb endpoint.proto yac.proto ordering.proto loader.proto
```
Comment (Warchant) [inline example/node/prepare.sh]: are you sure you need ordering, yac and loader?

File: shared_model/bindings/CMakeLists.txt [hunk 9]
```diff
@@ -73,3 +72,31 @@ if (SWIG_CSHARP)
     swig_link_libraries(libirohacs bindings)
     add_custom_target(irohacs DEPENDS ${SWIG_MODULE_libirohacs_REAL_NAME})
 endif()
+
+if (SWIG_NODE)
+
+    find_package (nodejs REQUIRED)
```
Comment (Warchant) [inline shared_model/bindings/CMakeLists.txt]: what if not found?

File: shared_model/bindings/CMakeLists.txt [hunk 10]
```diff
@@ -73,3 +72,31 @@ if (SWIG_CSHARP)
...
+        string (LENGTH "${V8_VERSION_HEX}" V8_VERSION_HEX_length)
+    endwhile ()
+
+    if (${CMAKE_SYSTEM_NAME} STREQUAL Darwin)
+        set(MAC_OPTS "-flat_namespace -undefined suppress")
+        set(MAC_SHARED_LIBRARY "SHARED")
+    endif()
+
+    set_property(SOURCE bindings.i PROPERTY SWIG_FLAGS "-node" "-DV8_VERSION=${V8_VERSION_HEX}")
+    message(STATUS "-I${CMAKE_BINARY_DIR}/../../../shared_model")
```
Comment (Warchant) [inline shared_model/bindings/CMakeLists.txt]: looks like a debug message

**Comment by luckychess** [comment]
Outdated. Nodejs support will be implemented in another pull request.

**Comment by luckychess** [comment]
Please do not touch this branch; it could be used in another prs.

**Pull request closed without merging**